drivers/spi/spi_mspm0.c

```diff
@@ -0,0 +1,296 @@
+
+	/* nop tx if no data is expected */
+	uint32_t tx_frame = 0;
+	if (spi_context_tx_buf_on(&data->ctx)) {
+		if (data->dfs == 1) {
+			tx_frame = UNALIGNED_GET((uint8_t *)(ctx->tx_buf));
+		} else if (data->dfs == 2) {
+			tx_frame = UNALIGNED_GET((uint16_t *)(ctx->tx_buf));
+		} else {
+			tx_frame = UNALIGNED_GET((uint32_t *)(ctx->tx_buf));
+		}
+	}
+
+	DL_SPI_transmitDataBlocking32(cfg->base, tx_frame);
+	while (DL_SPI_isBusy(cfg->base)) {
+		/* wait for tx fifo to be sent */
+	}
+
+	spi_context_update_tx(ctx, data->dfs, 1);
+}
+
+static void spi_mspm0_frame_rx(const struct device *dev)
+{
+	struct spi_mspm0_data *data = dev->data;
+	const struct spi_mspm0_config *cfg = dev->config;
+	struct spi_context *ctx = &data->ctx;
+
+	/* rx fifo will contain 0s if there is no data available, so make sure to clean out those */
+	const uint32_t rx_frame = DL_SPI_receiveDataBlocking32(cfg->base);
+	if (spi_context_rx_buf_on(&data->ctx)) {
+		if (data->dfs == 1) {
+			UNALIGNED_PUT(rx_frame, (uint8_t *)ctx->rx_buf);
+		} else if (data->dfs == 2) {
+			UNALIGNED_PUT(rx_frame, (uint16_t *)ctx->rx_buf);
+		} else {
+			UNALIGNED_PUT(rx_frame, (uint32_t *)ctx->rx_buf);
+		}
+
+		spi_context_update_rx(ctx, data->dfs, 1);
+	}
+}
+
+static void spi_mspm0_start_transfer(const struct device *dev)
+{
+	struct spi_mspm0_data *data = dev->data;
+
+	spi_context_cs_control(&data->ctx, true);
+
+	while (spi_mspm0_transfer_ongoing(data)) {
+		spi_mspm0_frame_tx(dev);
+		spi_mspm0_frame_rx(dev);
+	}
+
+	spi_context_cs_control(&data->ctx, false);
+	spi_context_complete(&data->ctx, dev, 0);
+}
+
+static int spi_mspm0_transceive(const struct device *dev, const struct spi_config *spi_cfg,
+				const struct spi_buf_set *tx_bufs,
+				const struct spi_buf_set *rx_bufs)
+{
+	struct spi_mspm0_data *data = dev->data;
+	spi_context_lock(&data->ctx, false, NULL, NULL, spi_cfg);
+
+	int ret = spi_mspm0_configure(dev, spi_cfg);
+	if (ret != 0) {
+		spi_context_release(&data->ctx, ret);
+		return ret;
+	}
+
+	spi_context_buffers_setup(&data->ctx, tx_bufs, rx_bufs, data->dfs);
+	spi_mspm0_start_transfer(dev);
+
+	ret = spi_context_wait_for_completion(&data->ctx);
+	spi_context_release(&data->ctx, ret);
+
+	return ret;
+}
+
+static int spi_mspm0_release(const struct device *dev, const struct spi_config *config)
+{
+
+	struct spi_mspm0_data *data = dev->data;
+	const struct spi_mspm0_config *cfg = dev->config;
+
+	if (!spi_context_configured(&data->ctx, config)) {
+		return -EINVAL;
+	}
+
+	if (DL_SPI_isBusy(cfg->base)) {
+		return -EBUSY;
+	}
+
+	spi_context_unlock_unconditionally(&data->ctx);
+	return 0;
+}
+
+static const struct spi_driver_api spi_mspm0_api = {
+	.transceive = spi_mspm0_transceive,
+	.release = spi_mspm0_release,
+};
+
+static int spi_mspm0_init(const struct device *dev)
+{
+	struct spi_mspm0_data *data = dev->data;
+	const struct spi_mspm0_config *cfg = dev->config;
+
+	DL_SPI_reset(cfg->base);
+	DL_SPI_enablePower(cfg->base);
+	delay_cycles(POWER_STARTUP_DELAY);
```

Comment:
Where does this `POWER_STARTUP_DELAY` come from? If it's an external define from TI HAL, maybe it makes sense to do a `#define MSPM0_SPI_STARTUP_DELAY POWER_STARTUP_DELAY` just to add a "local scope"-ish variant?

Reply:
It's a part of the `soc.h`. It's unfortunately required for their drivers. All the mspm0 driver behave like this. Would be nice to have it generalized but that's work for TI to implement.

Reply:
Okay, I see. Makes more sense to use this directly then.